File: Sources/CococoJinja/Expression/Parser.swift
//
//  Parser.swift
//  CococoJinja
//
//  Copyright Triple C Labs GmbH
//
//  Precedence-climbing parser: [Token] -> Expr. Precedence (lowest -> highest):
//  ternary -> or -> and -> not -> comparison/in -> ~ -> + - -> * / // % ->
//  unary - -> filter | -> member/index -> primary.
//  See docs/workflow-expression-dialect.md §4.
//
//  Member/index access is only valid from an identifier root (a reference); you
//  cannot index a literal or a filter result (`[1,2][0]`, `(x).y` do not parse).
//

import Foundation

enum Parser {
    /// Parse a complete expression. Throws if there are trailing tokens.
    static func parse(_ tokens: [Token]) throws -> Expr {
        var state = State(tokens)
        let expr = try state.parseExpression()
        guard state.peek.kind == .eof else {
            throw ExpressionError.parse(
                "unexpected \(state.peek.kind.description) after expression",
                at: state.peek.offset
            )
        }
        return expr
    }

    /// Convenience: lex + parse a source string.
    static func parse(source: String) throws -> Expr {
        try parse(Lexer.tokenize(source))
    }

    // MARK: - Parser state

    private struct State {
        private let tokens: [Token]
        private var pos = 0

        init(_ tokens: [Token]) {
            // `tokens` always ends in `.eof` (Lexer guarantees it).
            self.tokens = tokens
        }

        var peek: Token { tokens[pos] }
        private func peek(_ ahead: Int) -> Token {
            let i = pos + ahead
            return i < tokens.count ? tokens[i] : tokens[tokens.count - 1]
        }

        @discardableResult
        private mutating func advance() -> Token {
            defer { if pos < tokens.count - 1 { pos += 1 } }
            return tokens[pos]
        }

        private mutating func match(_ kind: TokenKind) -> Bool {
            guard peek.kind == kind else { return false }
            advance()
            return true
        }

        private mutating func expect(_ kind: TokenKind, _ context: String) throws {
            guard peek.kind == kind else {
                throw ExpressionError.parse(
                    "expected \(kind.description) \(context), found \(peek.kind.description)",
                    at: peek.offset
                )
            }
            advance()
        }

        // MARK: ternary (lowest)

        mutating func parseExpression() throws -> Expr {
            let value = try parseOr()
            guard match(.kwIf) else { return value }
            let condition = try parseOr()
            try expect(.kwElse, "in conditional expression")
            let otherwise = try parseExpression() // right-associative
            return .conditional(condition: condition, then: value, otherwise: otherwise)
        }

        // MARK: or / and (n-ary)

        private mutating func parseOr() throws -> Expr {
            var operands = [try parseAnd()]
            while match(.kwOr) { operands.append(try parseAnd()) }
            return operands.count == 1 ? operands[0] : .logical(.or, operands)
        }

        private mutating func parseAnd() throws -> Expr {
            var operands = [try parseNot()]
            while match(.kwAnd) { operands.append(try parseNot()) }
            return operands.count == 1 ? operands[0] : .logical(.and, operands)
        }

        // MARK: not (prefix, looser than comparison)

        private mutating func parseNot() throws -> Expr {
            if match(.kwNot) {
                return .unary(.not, try parseNot())
            }
            return try parseComparison()
        }

        // MARK: comparison & membership (left-associative)

        private mutating func parseComparison() throws -> Expr {
            var left = try parseConcat()
            loop: while true {
                let op: BinaryOperator
                switch peek.kind {
                case .eq: op = .eq
                case .neq: op = .neq
                case .lt: op = .lt
                case .lte: op = .lte
                case .gt: op = .gt
                case .gte: op = .gte
                case .kwIn: op = .contains
                case .kwNot where peek(1).kind == .kwIn:
                    advance() // not
                    advance() // in
                    left = .binary(.notContains, left, try parseConcat())
                    continue loop
                default:
                    break loop
                }
                advance()
                left = .binary(op, left, try parseConcat())
            }
            return left
        }

        // MARK: ~ concat (left-associative)

        private mutating func parseConcat() throws -> Expr {
            var left = try parseAdditive()
            while match(.tilde) {
                left = .binary(.concat, left, try parseAdditive())
            }
            return left
        }

        // MARK: additive (left-associative)

        private mutating func parseAdditive() throws -> Expr {
            var left = try parseMultiplicative()
            while true {
                let op: BinaryOperator
                switch peek.kind {
                case .plus: op = .add
                case .minus: op = .subtract
                default: return left
                }
                advance()
                left = .binary(op, left, try parseMultiplicative())
            }
        }

        // MARK: multiplicative (left-associative)

        private mutating func parseMultiplicative() throws -> Expr {
            var left = try parseUnary()
            while true {
                let op: BinaryOperator
                switch peek.kind {
                case .star: op = .multiply
                case .slash: op = .divide
                case .slashSlash: op = .floorDivide
                case .percent: op = .modulo
                default: return left
                }
                advance()
                left = .binary(op, left, try parseUnary())
            }
        }

        // MARK: unary - / + (prefix)

        private mutating func parseUnary() throws -> Expr {
            if match(.plus) {
                return try parseUnary() // unary plus is a no-op
            }
            if match(.minus) {
                let operand = try parseUnary()
                // Constant-fold negation of a numeric literal so `-5` stays a bare
                // literal (keeps it inside the builder's rule subset, §5).
                switch operand {
                case let .literal(.int(n)): return .literal(.int(-n))
                case let .literal(.double(d)): return .literal(.double(-d))
                default: return .unary(.negate, operand)
                }
            }
            return try parseFilter()
        }

        // MARK: filter | (left-associative)

        private mutating func parseFilter() throws -> Expr {
            var input = try parsePostfix()
            while match(.pipe) {
                guard case let .identifier(name) = peek.kind else {
                    throw ExpressionError.parse(
                        "expected filter name after '|', found \(peek.kind.description)",
                        at: peek.offset
                    )
                }
                advance()
                var args: [Expr] = []
                if match(.lparen) {
                    args = try parseCommaSeparated(until: .rparen)
                    try expect(.rparen, "after filter arguments")
                }
                input = .filter(name: name, input: input, arguments: args)
            }
            return input
        }

        // MARK: postfix member/index (only on identifier-rooted references)

        private mutating func parsePostfix() throws -> Expr {
            // Only an identifier starts a reference with member/index access.
            guard case let .identifier(root) = peek.kind else {
                return try parsePrimary()
            }
            advance()
            var segments: [PathSegment] = [.key(root)]
            loop: while true {
                switch peek.kind {
                case .dot:
                    advance()
                    guard case let .identifier(name) = peek.kind else {
                        throw ExpressionError.parse(
                            "expected member name after '.', found \(peek.kind.description)",
                            at: peek.offset
                        )
                    }
                    advance()
                    segments.append(.key(name))
                case .lbracket:
                    advance()
                    segments.append(try parseSubscript())
                    try expect(.rbracket, "after subscript")
                default:
                    break loop
                }
            }
            return .reference(segments)
        }

        /// Parse the contents of `[ ... ]` into a path segment.
        private mutating func parseSubscript() throws -> PathSegment {
            // Literal int → index; literal string → key; anything else → dynamic.
            switch peek.kind {
            case let .int(n) where peek(1).kind == .rbracket:
                advance()
                return .index(n)
            case let .string(s) where peek(1).kind == .rbracket:
                advance()
                return .key(s)
            default:
                return .dynamic(try parseExpression())
            }
        }

        // MARK: primary

        private mutating func parsePrimary() throws -> Expr {
            let token = peek
            switch token.kind {
            case let .int(v):
                advance(); return .literal(.int(v))
            case let .double(v):
                advance(); return .literal(.double(v))
            case let .string(v):
                advance(); return .literal(.string(v))
            case .kwTrue:
                advance(); return .literal(.bool(true))
            case .kwFalse:
                advance(); return .literal(.bool(false))
            case .kwNull:
                advance(); return .literal(.null)
            case .identifier:
                // Reached only when not handled by parsePostfix (it always is),
                // but kept for completeness.
                return try parsePostfix()
            case .lparen:
                advance()
                let inner = try parseExpression()
                try expect(.rparen, "after parenthesized expression")
                return inner
            case .lbracket:
                advance()
                let elements = try parseCommaSeparated(until: .rbracket)
                try expect(.rbracket, "after array literal")
                return .arrayLiteral(elements)
            case .lbrace:
                return try parseObjectLiteral()
            default:
                throw ExpressionError.parse(
                    "unexpected \(token.kind.description)",
                    at: token.offset
                )
            }
        }

        private mutating func parseObjectLiteral() throws -> Expr {
            try expect(.lbrace, "to open object literal")
            var entries: [ObjectEntry] = []
            if peek.kind != .rbrace {
                while true {
                    let key: String
                    switch peek.kind {
                    case let .string(s): key = s; advance()
                    case let .identifier(name): key = name; advance()
                    default:
                        throw ExpressionError.parse(
                            "expected object key (string or identifier), found \(peek.kind.description)",
                            at: peek.offset
                        )
                    }
                    try expect(.colon, "after object key")
                    let value = try parseExpression()
                    entries.append(ObjectEntry(key: key, value: value))
                    if match(.comma) {
                        if peek.kind == .rbrace { break } // trailing comma
                        continue
                    }
                    break
                }
            }
            try expect(.rbrace, "after object literal")
            return .objectLiteral(entries)
        }

        /// Parse zero or more comma-separated expressions, stopping at `terminator`
        /// (which is NOT consumed). Allows a single trailing comma.
        private mutating func parseCommaSeparated(until terminator: TokenKind) throws -> [Expr] {
            var items: [Expr] = []
            if peek.kind == terminator { return items }
            while true {
                items.append(try parseExpression())
                if match(.comma) {
                    if peek.kind == terminator { break } // trailing comma
                    continue
                }
                break
            }
            return items
        }
    }
}

File: ts/src/parser.ts
// parser.ts — port of Expression/Parser.swift
//
// Precedence-climbing parser: Token[] -> Expr. Precedence (lowest -> highest):
// ternary -> or -> and -> not -> comparison/in -> ~ -> + - -> * / // % ->
// unary - -> filter | -> member/index -> primary.

import type {
	BinaryOperator,
	Expr,
	ObjectEntry,
	PathSegment,
	Token,
	TokenKind,
} from "./ast.ts";
import { tokenDescription } from "./ast.ts";
import { ExpressionError } from "./errors.ts";
import { tokenize } from "./lexer.ts";
import { JV } from "./value.ts";

export function parse(tokens: Token[]): Expr {
	const state = new State(tokens);
	const expr = state.parseExpression();
	if (state.peek().kind.t !== "eof") {
		throw ExpressionError.parse(
			`unexpected ${tokenDescription(state.peek().kind)} after expression`,
			state.peek().offset,
		);
	}
	return expr;
}

/** Convenience: lex + parse a source string. */
export function parseSource(source: string): Expr {
	return parse(tokenize(source));
}

class State {
	private tokens: Token[];
	private pos = 0;

	constructor(tokens: Token[]) {
		this.tokens = tokens;
	}

	peek(): Token {
		return this.tokens[this.pos]!;
	}
	private peekAhead(ahead: number): Token {
		const i = this.pos + ahead;
		return i < this.tokens.length ? this.tokens[i]! : this.tokens[this.tokens.length - 1]!;
	}
	private advance(): Token {
		const t = this.tokens[this.pos]!;
		if (this.pos < this.tokens.length - 1) this.pos += 1;
		return t;
	}
	private matchKind(t: TokenKind["t"]): boolean {
		if (this.peek().kind.t === t) {
			this.advance();
			return true;
		}
		return false;
	}
	private expect(t: TokenKind["t"], context: string): void {
		if (this.peek().kind.t !== t) {
			throw ExpressionError.parse(
				`expected ${describeKind(t)} ${context}, found ${tokenDescription(this.peek().kind)}`,
				this.peek().offset,
			);
		}
		this.advance();
	}

	// MARK: ternary (lowest)

	parseExpression(): Expr {
		const value = this.parseOr();
		if (!this.matchKind("kwIf")) return value;
		const condition = this.parseOr();
		this.expect("kwElse", "in conditional expression");
		const otherwise = this.parseExpression(); // right-associative
		return { e: "conditional", condition, then: value, otherwise };
	}

	// MARK: or / and (n-ary)

	private parseOr(): Expr {
		const operands = [this.parseAnd()];
		while (this.matchKind("kwOr")) operands.push(this.parseAnd());
		return operands.length === 1 ? operands[0]! : { e: "logical", op: "or", operands };
	}

	private parseAnd(): Expr {
		const operands = [this.parseNot()];
		while (this.matchKind("kwAnd")) operands.push(this.parseNot());
		return operands.length === 1 ? operands[0]! : { e: "logical", op: "and", operands };
	}

	// MARK: not (prefix, looser than comparison)

	private parseNot(): Expr {
		if (this.matchKind("kwNot")) {
			return { e: "unary", op: "not", operand: this.parseNot() };
		}
		return this.parseComparison();
	}

	// MARK: comparison & membership (left-associative)

	private parseComparison(): Expr {
		let left = this.parseConcat();
		loop: while (true) {
			let op: BinaryOperator;
			switch (this.peek().kind.t) {
				case "eq":
					op = "==";
					break;
				case "neq":
					op = "!=";
					break;
				case "lt":
					op = "<";
					break;
				case "lte":
					op = "<=";
					break;
				case "gt":
					op = ">";
					break;
				case "gte":
					op = ">=";
					break;
				case "kwIn":
					op = "in";
					break;
				case "kwNot":
					if (this.peekAhead(1).kind.t === "kwIn") {
						this.advance(); // not
						this.advance(); // in
						left = { e: "binary", op: "not in", lhs: left, rhs: this.parseConcat() };
						continue loop;
					}
					break loop;
				default:
					break loop;
			}
			this.advance();
			left = { e: "binary", op, lhs: left, rhs: this.parseConcat() };
		}
		return left;
	}

	// MARK: ~ concat (left-associative)

	private parseConcat(): Expr {
		let left = this.parseAdditive();
		while (this.matchKind("tilde")) {
			left = { e: "binary", op: "~", lhs: left, rhs: this.parseAdditive() };
		}
		return left;
	}

	// MARK: additive (left-associative)

	private parseAdditive(): Expr {
		let left = this.parseMultiplicative();
		while (true) {
			let op: BinaryOperator;
			switch (this.peek().kind.t) {
				case "plus":
					op = "+";
					break;
				case "minus":
					op = "-";
					break;
				default:
					return left;
			}
			this.advance();
			left = { e: "binary", op, lhs: left, rhs: this.parseMultiplicative() };
		}
	}

	// MARK: multiplicative (left-associative)

	private parseMultiplicative(): Expr {
		let left = this.parseUnary();
		while (true) {
			let op: BinaryOperator;
			switch (this.peek().kind.t) {
				case "star":
					op = "*";
					break;
				case "slash":
					op = "/";
					break;
				case "slashSlash":
					op = "//";
					break;
				case "percent":
					op = "%";
					break;
				default:
					return left;
			}
			this.advance();
			left = { e: "binary", op, lhs: left, rhs: this.parseUnary() };
		}
	}

	// MARK: unary - / + (prefix)

	private parseUnary(): Expr {
		if (this.matchKind("plus")) {
			return this.parseUnary(); // unary plus is a no-op
		}
		if (this.matchKind("minus")) {
			const operand = this.parseUnary();
			// Constant-fold negation of a numeric literal.
			if (operand.e === "literal") {
				if (operand.value.kind === "int") {
					return { e: "literal", value: JV.int(-operand.value.value) };
				}
				if (operand.value.kind === "double") {
					return { e: "literal", value: JV.double(-operand.value.value) };
				}
			}
			return { e: "unary", op: "negate", operand };
		}
		return this.parseFilter();
	}

	// MARK: filter | (left-associative)

	private parseFilter(): Expr {
		let input = this.parsePostfix();
		while (this.matchKind("pipe")) {
			const k = this.peek().kind;
			if (k.t !== "identifier") {
				throw ExpressionError.parse(
					`expected filter name after '|', found ${tokenDescription(k)}`,
					this.peek().offset,
				);
			}
			const name = k.value;
			this.advance();
			let args: Expr[] = [];
			if (this.matchKind("lparen")) {
				args = this.parseCommaSeparated("rparen");
				this.expect("rparen", "after filter arguments");
			}
			input = { e: "filter", name, input, arguments: args };
		}
		return input;
	}

	// MARK: postfix member/index (only on identifier-rooted references)

	private parsePostfix(): Expr {
		const rootKind = this.peek().kind;
		if (rootKind.t !== "identifier") {
			return this.parsePrimary();
		}
		this.advance();
		const segments: PathSegment[] = [{ s: "key", name: rootKind.value }];
		loop: while (true) {
			switch (this.peek().kind.t) {
				case "dot": {
					this.advance();
					const nameKind = this.peek().kind;
					if (nameKind.t !== "identifier") {
						throw ExpressionError.parse(
							`expected member name after '.', found ${tokenDescription(nameKind)}`,
							this.peek().offset,
						);
					}
					this.advance();
					segments.push({ s: "key", name: nameKind.value });
					break;
				}
				case "lbracket": {
					this.advance();
					segments.push(this.parseSubscript());
					this.expect("rbracket", "after subscript");
					break;
				}
				default:
					break loop;
			}
		}
		return { e: "reference", segments };
	}

	/** Parse the contents of `[ ... ]` into a path segment. */
	private parseSubscript(): PathSegment {
		const k = this.peek().kind;
		if (k.t === "int" && this.peekAhead(1).kind.t === "rbracket") {
			this.advance();
			return { s: "index", index: k.value };
		}
		if (k.t === "string" && this.peekAhead(1).kind.t === "rbracket") {
			this.advance();
			return { s: "key", name: k.value };
		}
		return { s: "dynamic", expr: this.parseExpression() };
	}

	// MARK: primary

	private parsePrimary(): Expr {
		const token = this.peek();
		const k = token.kind;
		switch (k.t) {
			case "int":
				this.advance();
				return { e: "literal", value: JV.int(k.value) };
			case "double":
				this.advance();
				return { e: "literal", value: JV.double(k.value) };
			case "string":
				this.advance();
				return { e: "literal", value: JV.string(k.value) };
			case "kwTrue":
				this.advance();
				return { e: "literal", value: JV.bool(true) };
			case "kwFalse":
				this.advance();
				return { e: "literal", value: JV.bool(false) };
			case "kwNull":
				this.advance();
				return { e: "literal", value: JV.null };
			case "identifier":
				return this.parsePostfix();
			case "lparen": {
				this.advance();
				const inner = this.parseExpression();
				this.expect("rparen", "after parenthesized expression");
				return inner;
			}
			case "lbracket": {
				this.advance();
				const elements = this.parseCommaSeparated("rbracket");
				this.expect("rbracket", "after array literal");
				return { e: "arrayLiteral", elements };
			}
			case "lbrace":
				return this.parseObjectLiteral();
			default:
				throw ExpressionError.parse(`unexpected ${tokenDescription(k)}`, token.offset);
		}
	}

	private parseObjectLiteral(): Expr {
		this.expect("lbrace", "to open object literal");
		const entries: ObjectEntry[] = [];
		if (this.peek().kind.t !== "rbrace") {
			while (true) {
				const kk = this.peek().kind;
				let key: string;
				if (kk.t === "string") {
					key = kk.value;
					this.advance();
				} else if (kk.t === "identifier") {
					key = kk.value;
					this.advance();
				} else {
					throw ExpressionError.parse(
						`expected object key (string or identifier), found ${tokenDescription(kk)}`,
						this.peek().offset,
					);
				}
				this.expect("colon", "after object key");
				const value = this.parseExpression();
				entries.push({ key, value });
				if (this.matchKind("comma")) {
					if (this.peek().kind.t === "rbrace") break; // trailing comma
					continue;
				}
				break;
			}
		}
		this.expect("rbrace", "after object literal");
		return { e: "objectLiteral", entries };
	}

	/** Zero or more comma-separated expressions until `terminator` (not consumed). */
	private parseCommaSeparated(terminator: TokenKind["t"]): Expr[] {
		const items: Expr[] = [];
		if (this.peek().kind.t === terminator) return items;
		while (true) {
			items.push(this.parseExpression());
			if (this.matchKind("comma")) {
				if (this.peek().kind.t === terminator) break; // trailing comma
				continue;
			}
			break;
		}
		return items;
	}
}

// Describe a bare token-kind tag for `expect` messages, matching Swift's
// `TokenKind.description` for the punctuation/keyword kinds it is used with.
function describeKind(t: TokenKind["t"]): string {
	const map: Partial<Record<TokenKind["t"], string>> = {
		rparen: "')'",
		rbracket: "']'",
		rbrace: "'}'",
		lbrace: "'{'",
		colon: "':'",
		kwElse: "'else'",
	};
	return map[t] ?? `'${t}'`;
}
